relay/plugins/traffic/content-blocker-plugin/content-blocker-plugin.go

```diff
@@ -0,0 +1,296 @@
+package content_blocker_plugin
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+
+	"github.com/fullstorydev/relay-core/relay/plugins/traffic"
+)
+
+var (
+	Factory    contentBlockerPluginFactory
+	logger     = log.New(os.Stdout, "[traffic-content-blocker] ", 0)
+	pluginName = "Content-Blocker"
+
+	excludeBodyContentVar   = "TRAFFIC_EXCLUDE_BODY_CONTENT"   // A go regexp string or empty
+	maskBodyContentVar      = "TRAFFIC_MASK_BODY_CONTENT"      // A go regexp string or empty
+	excludeHeaderContentVar = "TRAFFIC_EXCLUDE_HEADER_CONTENT" // A go regexp string or empty
+	maskHeaderContentVar    = "TRAFFIC_MASK_HEADER_CONTENT"    // A go regexp string or empty
+)
+
+type contentBlockerPluginFactory struct{}
+
+func (f contentBlockerPluginFactory) Name() string {
+	return pluginName
+}
+
+func (f contentBlockerPluginFactory) ConfigVars() map[string]bool {
+	return map[string]bool{
+		excludeBodyContentVar:   false,
+		maskBodyContentVar:      false,
+		excludeHeaderContentVar: false,
+		maskHeaderContentVar:    false,
+	}
+}
+
+func (f contentBlockerPluginFactory) New(env map[string]string) (traffic.Plugin, error) {
+	bodyBlockers, err := newContentBlockerList(
+		env,
+		"body",
+		excludeBodyContentVar,
+		maskBodyContentVar,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	headerBlockers, err := newContentBlockerList(
+		env,
+		"header",
+		excludeHeaderContentVar,
+		maskHeaderContentVar,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(bodyBlockers) == 0 && len(headerBlockers) == 0 {
+		return nil, nil
+	}
+
+	return &contentBlockerPlugin{
+		bodyBlockers:   bodyBlockers,
+		headerBlockers: headerBlockers,
+	}, nil
+}
+
+type contentBlockerPlugin struct {
+	bodyBlockers   []*contentBlocker
+	headerBlockers []*contentBlocker
+}
+
+func (plug contentBlockerPlugin) Name() string {
+	return pluginName
+}
+
+func (plug contentBlockerPlugin) HandleRequest(response http.ResponseWriter, request *http.Request, serviced bool) bool {
+	if serviced := plug.blockHeaderContent(response, request); serviced {
+		return true
+	}
+	if serviced := plug.blockBodyContent(response, request); serviced {
+		return true
+	}
+	return false
+}
+
+func (plug contentBlockerPlugin) blockHeaderContent(response http.ResponseWriter, request *http.Request) bool {
+	if len(plug.headerBlockers) == 0 {
+		return false
+	}
+
+	for _, headerValues := range request.Header {
+		for i, headerValue := range headerValues {
+			processedValue := []byte(headerValue)
+			for _, blocker := range plug.headerBlockers {
+				processedValue = blocker.Block(processedValue)
+			}
+			headerValues[i] = string(processedValue)
+		}
+	}
+
+	return false
+}
+
+func (plug contentBlockerPlugin) blockBodyContent(response http.ResponseWriter, request *http.Request) bool {
+	if len(plug.bodyBlockers) == 0 {
+		return false
+	}
+
+	// Block all websocket connections if we're blocking body content.
```

Comment:
This is okay but may need communication with the rest of the team so we're all aware.  CX may also need to be aware.

Reply:
There are no overlap between relay users + websocket users (well no websocket users) so more of a thing to be aware of in the future.

Reply:
+1. Clint and I discussed this as a short term solution, but I'll make sure it gets wider visibility within the team. We've got a proper fix on the backlog, although we probably can't get to it right away - basically, we need to rearchitect the plugin interface a bit so that there's a hook for websocket content analogous to the hook for HTTP content.